File: Evo Island/cpp/agent.cpp
#include "agent.h"
#include <algorithm>

void Agent::decode_genome() {
    lifespan               = genome[0];
    hardiness              = genome[1];
    strength               = genome[2];
    metabolism             = genome[3];
    reproduction_threshold = genome[4];
    speed                  = genome[5];
    trophism        = genome[6] / 100.0f;           // [0,100] → [0,1]
    kin_attraction  = (genome[7] - 50.0f) / 50.0f; // [0,100] → [-1,1]
    threat_response = (genome[8] - 50.0f) / 50.0f; // [0,100] → [-1,1]
}

void Agent::calculate_genetic_distance() {
    float sum = 0.0f;
    for (int i = 0; i < 9; ++i) {
        float d = genome[i] - ANCESTOR[i];
        sum += d * d;
    }
    genetic_distance = std::sqrt(sum);
}

void Agent::kill() {
    alive                  = false;
    age                    = 0.0f;
    energy_reserves        = 0.0f;
    genetic_distance       = 0.0f;
    species                = 0;
    lifespan               = 0.0f;
    hardiness              = 0.0f;
    strength               = 0.0f;
    metabolism             = 0.0f;
    reproduction_threshold = 0.0f;
    speed                  = 0.0f;
    trophism        = 0.0f;
    kin_attraction  = 0.0f;
    threat_response = 0.0f;
    for (int i = 0; i < 9; ++i) genome[i] = 0.0f;
    for (int i = 0; i < 3; ++i) color[i]  = 0.0f;
}

// Death probability follows a logistic curve centred at lifespan/2.
// steepness=10 produces a sharp S — near-zero early, near-certain past midpoint.
void Agent::age_step(bool enable_aging, std::mt19937& rng) {
    age += 1.0f;
    if (!enable_aging) return;

    float midpoint = lifespan / 2.0f;
    if (midpoint == 0.0f) { kill(); return; }

    constexpr float steepness = 10.0f;
    float death_prob = 1.0f / (1.0f + std::exp(
        -steepness * (age - midpoint) / midpoint));

    std::uniform_real_distribution<float> uni(0.0f, 1.0f);
    if (uni(rng) < death_prob)
        kill();
}

void Agent::consume_food(float f) { energy_reserves += f; }
void Agent::metabolize(float e)   { energy_reserves -= e; }

// Map all 9 genome values to RGB via a fixed random projection.
// Each gene is normalised to [0,1] and centred on the ANCESTOR so the
// initial population sits near the middle of the colour cube.  The fixed
// matrix means the same genome always produces the same colour — genetic
// neighbours share similar colours regardless of which genes drifted.
void Agent::genome_to_color(const float g[9], float c[3]) {
    // Fixed 9×3 projection matrix — entries sampled offline from N(0,1).
    // Never change these: doing so would recolour every existing result.
    static constexpr float W[9][3] = {
        { 0.4122f, -0.8090f,  0.4226f},
        {-0.6577f,  0.1919f,  0.6122f},
        { 0.5062f,  0.5530f, -0.2150f},
        {-0.2348f, -0.4685f, -0.7312f},
        { 0.7819f, -0.1234f,  0.3456f},
        {-0.3291f,  0.6782f, -0.5521f},
        { 0.1456f,  0.3891f,  0.8012f},
        {-0.5678f, -0.2345f,  0.1234f},
        { 0.6789f,  0.4512f, -0.3901f},
    };
    // ANCESTOR normalised to [0,1] — centres projection so initial population
    // maps to ~(0.5, 0.5, 0.5) and diverges outward as evolution proceeds.
    static constexpr float ANORM[9] = {
        0.20f, 0.10f, 0.05f, 0.05f, 0.03f, 0.00f, 0.00f, 0.50f, 0.50f
    };
    constexpr float SCALE = 3.0f;
    for (int j = 0; j < 3; ++j) {
        float y = 0.0f;
        for (int i = 0; i < 9; ++i)
            y += W[i][j] * (g[i] / 100.0f - ANORM[i]);
        c[j] = 1.0f / (1.0f + std::exp(-y * SCALE));
    }
}

Agent Agent::create_live(const float g[9]) {
    Agent a;
    a.alive = true;
    for (int i = 0; i < 9; ++i) a.genome[i] = g[i];
    genome_to_color(a.genome, a.color);
    a.decode_genome();
    a.calculate_genetic_distance();
    a.energy_reserves = 5.0f;
    return a;
}

Agent Agent::create_live_default() {
    return create_live(ANCESTOR);
}

// Each gene mutates independently with probability mutation_rate;
// effect is sampled from N(0, 2) and clamped to [0, 100].
void Agent::mutate_genome(const float in[9], float out[9],
                          float mutation_rate, std::mt19937& rng) {
    std::uniform_real_distribution<float> uni(0.0f, 1.0f);
    std::normal_distribution<float>       norm_dist(0.0f, 2.0f);
    for (int i = 0; i < 9; ++i) {
        float effect = (uni(rng) < mutation_rate) ? norm_dist(rng) : 0.0f;
        out[i] = std::min(100.0f, std::max(0.0f, in[i] + effect));
    }
}

Agent Agent::reproduce_asexually(const Agent& parent,
                                  float mutation_rate, std::mt19937& rng) {
    float child_genome[9];
    mutate_genome(parent.genome, child_genome, mutation_rate, rng);
    return create_live(child_genome);
}

File: Evo Island/py/agent.py
import math
import random

import matplotlib.colors as mcolors
import numpy as np


class Agent:
    # Genome: [lifespan, hardiness, strength, metabolism, reproduction_threshold, speed]
    # speed=0 gives sessile (plant-like) behaviour as the evolutionary baseline.
    common_ancestor_genome = np.array([20, 10, 5, 5, 3, 0])

    def __init__(self):
        self.alive = False
        self.genome = None
        self.color = None
        self.age = 0
        self.energy_reserves = 0
        self.genetic_distance = None
        self.species = 0
        self.reset_traits()

    @classmethod
    def create_live_agent(cls, genome=None):
        live_agent = cls()
        live_agent.alive = True
        live_agent.genome = (
            genome if genome is not None else cls.generate_default_genome()
        )
        live_agent.color = cls.genome_to_color(live_agent.genome)
        live_agent.decode_genome()
        live_agent.calculate_genetic_distance()
        live_agent.energy_reserves = 5
        return live_agent

    def decode_genome(self):
        if self.genome is not None:
            self.lifespan = self.genome[0]
            self.hardiness = self.genome[1]
            self.strength = self.genome[2]
            self.metabolism = self.genome[3]
            self.reproduction_threshold = self.genome[4]
            self.speed = self.genome[5]

    @staticmethod
    def generate_default_genome():
        return Agent.common_ancestor_genome.copy()

    @staticmethod
    def mutate_genome(genome, mutation_rate):
        """Per-gene Gaussian mutations, clamped to [0, 100]."""
        mutation_effects = np.zeros_like(genome)
        for i in range(len(genome)):
            if np.random.rand() < mutation_rate:
                mutation_effects[i] = np.random.normal(loc=0, scale=2)
        return np.clip(genome + mutation_effects, 0, 100)

    def calculate_genetic_distance(self, other_genome=None):
        if other_genome is None:
            other_genome = Agent.common_ancestor_genome
        if self.genome is not None:
            self.genetic_distance = np.linalg.norm(self.genome - other_genome)
        return self.genetic_distance

    @classmethod
    def reproduce_asexually(cls, parent_agent, mutation_rate):
        child_genome = cls.mutate_genome(parent_agent.genome, mutation_rate)
        return cls.create_live_agent(child_genome)

    def age_agent(self, enable_aging):
        """Logistic death curve centred at lifespan/2."""
        self.age += 1
        if enable_aging:
            midpoint = self.lifespan / 2
            if midpoint == 0:
                self.kill_agent()
                return
            steepness = 10
            death_probability = 1 / (
                1 + math.exp(-steepness * (self.age - midpoint) / midpoint)
            )
            if random.random() < death_probability:
                self.kill_agent()

    def consume_food(self, food_consumed):
        self.energy_reserves += food_consumed

    def metabolize(self, energy_used):
        self.energy_reserves -= energy_used

    def reset_traits(self):
        self.age = 0
        self.energy_reserves = 0
        self.lifespan = 0
        self.hardiness = 0
        self.strength = 0
        self.metabolism = 0
        self.reproduction_threshold = 0
        self.speed = 0
        self.genetic_distance = 0
        self.species = 0

    def kill_agent(self):
        self.alive = False
        self.genome = None
        self.color = None
        self.reset_traits()

    @staticmethod
    def genome_to_color(genome):
        """Map genome to RGB via HSV. Related genomes share hues for visual ID.

        Normalises values relative to the genome's own min/max so colour
        encodes internal shape rather than absolute scale.
        """
        genome_range = np.max(genome) - np.min(genome)
        norm_genome = (genome - np.min(genome)) / (genome_range if genome_range != 0 else 1)

        hue        = (norm_genome[0] * 360) % 360
        saturation = 50 + norm_genome[1] * 50  # [50%, 100%] — avoids washed-out greys
        lightness  = 50 + norm_genome[2] * 50  # [50%, 100%] — avoids black cells

        return mcolors.hsv_to_rgb((hue / 360, saturation / 100, lightness / 100))
